### PR TITLE
Feat/provider interface

### PR DESCRIPTION
**Pull request by shanejonas**
BREAKING

this will bump to `2.0.0` since it breaks the `.request` interface

changes from:

```
const result = await client.request("addition", [2, 2]});
```

to:

```
const result = await client.request({method: "addition", params: [2, 2]});
```

Also fixes up the typedoc:

image

### CODE REVIEW

File: src/Error.ts [hunk 1]
```diff
@@ -1,8 +1,10 @@
+import { ProviderRpcError } from "./ProviderInterface";
+
 export const ERR_TIMEOUT = 7777;
 export const ERR_MISSIING_ID = 7878;
 export const ERR_UNKNOWN = 7979;
 
-export class JSONRPCError extends Error {
+export class JSONRPCError extends Error implements ProviderRpcError{
```
Comment (zcstarr) [inline src/Error.ts:7]: why not just refactor this class to support the new type explictly ?

File: src/ProviderInterface.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+export interface ProviderRequestArguments {
```
Comment (zcstarr) [inline src/ProviderInterface.ts:1]: Maybe just RequestArguments, and maybe src/RequestInterface.ts

File: src/ProviderInterface.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  readonly method: string;
+  readonly params?: readonly unknown[] | object;
+}
+
+export interface Provider {
```
Comment (zcstarr) [inline src/ProviderInterface.ts:6]: Maybe IClient or ClientRequest

File: src/RequestManager.ts [hunk 4]
```diff
@@ -16,7 +17,7 @@ export interface IRequestEvents {
  * If a transport fails, or times out, move on to the next.
  */
 
-class RequestManager {
+class RequestManager implements Provider {
```
Comment (zcstarr) [inline src/RequestManager.ts:20]: as mentioned before implements Request ?

File: src/RequestManager.ts [hunk 5]
```diff
@@ -43,11 +44,11 @@ class RequestManager {
     return this.transports[0];
   }
 
-  public async request(method: string, params: any[], notification: boolean = false, timeout?: number): Promise<any> {
+  public async request(args: ProviderRequestArguments, notification: boolean = false, timeout?: number): Promise<any> {
```
Comment (zcstarr) [inline src/RequestManager.ts:47]: RequestArguments

File: src/index.ts [hunk 6]
```diff
@@ -19,12 +16,12 @@ interface IClient {
...
  * const transport = new HTTPTransport('http://localhost:3333');
  * const client = new Client(new RequestManager([transport]));
- * const result = await client.request(‘addition’, [2, 2]);
+ * const result = await client.request({method: 'addition', params: [2, 2]});
  * // => { jsonrpc: '2.0', id: 1, result: 4 }
  * ```
  *
  */
-class Client implements IClient {
+class Client implements Provider {
```
Comment (zcstarr) [inline src/index.ts:24]: implements Request

File: src/index.ts [hunk 7]
```diff
@@ -66,19 +63,22 @@ class Client implements IClient {
...
    * followed by a period character (U+002E or ASCII 46) are reserved for rpc-internal methods and extensions and
    * MUST NOT be used for anything else.
-   * @param params A Structured value that holds the parameter values to be used during the invocation of the method.
+   * @param requestObject.params A Structured value that holds the parameter values to be used during the invocation of the method.
+   *
+   * @example
+   * myClient.request({method: "foo", params: ["bar"]}).then(() => console.log('foobar'));
    */
-  public async request(method: string, params: any, timeout?: number) {
+  public async request(requestObject: ProviderRequestArguments, timeout?: number) {
```
Comment (zcstarr) [inline src/index.ts:74]: ditto

File: src/index.ts [hunk 8]
```diff
@@ -66,19 +63,22 @@ class Client implements IClient {
...
     await this.requestManager.connectPromise;
-    return this.requestManager.request(method, params, false, timeout);
+    return this.requestManager.request(requestObject, false, timeout);
   }
 
-  public async notify(method: string, params: any) {
+  public async notify(requestObject: ProviderRequestArguments) {
```
Comment (zcstarr) [inline src/index.ts:79]: ditto

**Comment by shanejonas** [comment]
need to fix the `.notify` method interface as well

File: src/Error.ts [hunk 9]
```diff
@@ -1,8 +1,10 @@
+import { JSONRpcError } from "./ClientInterface";
+
 export const ERR_TIMEOUT = 7777;
 export const ERR_MISSIING_ID = 7878;
 export const ERR_UNKNOWN = 7979;
 
-export class JSONRPCError extends Error {
+export class JSONRPCError extends Error implements JSONRpcError {
```
Comment (zcstarr) [inline src/Error.ts:7]: why not implement this directly changing the type data?:unknown

File: src/RequestManager.ts [hunk 10]
```diff
@@ -43,11 +44,11 @@ class RequestManager {
     return this.transports[0];
   }
 
-  public async request(method: string, params: any[], notification: boolean = false, timeout?: number): Promise<any> {
+  public async request(requestObject: RequestArguments, notification: boolean = false, timeout?: number): Promise<any> {
```
Comment (zcstarr) [inline src/RequestManager.ts:47]: In theory thew new Message type should be the type here, I think it might read better

File: package.json [hunk 11]
```diff
@@ -4,7 +4,7 @@
   "description": " A browser-compatible JSON-RPC client with multiple transports.",
   "main": "build/index.js",
   "scripts": {
-    "build": "tsc && typedoc --out docs",
+    "build": "tsc && typedoc --out docs --includeDeclarations --excludeExternals  --excludeNotExported --excludePrivate --hideGenerator --mode file --readme none",
```
Comment (BelfordZ) [inline package.json:7]: should use typedoc config file

**Pull request merged**